2023-09-24 00:29:59,718 - __main__ - INFO - Starting exam extraction...
2023-09-24 00:29:59,718 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 00:29:59,723 - root - DEBUG - ConfigManager.__new__ config_path: C:\Users\Tariq\Documents\birkbeck-project\config\config.yaml
2023-09-24 00:29:59,732 - root - DEBUG - self._config in ConfigManager: {'exam_formats': {'cambridge_science': {'imageProcessor': {'binary_threshold': 180, 'margin_start': 0, 'margin_end': 180, 'question_x_start': '', 'question_x_end': '', 'question_y_start': '', 'footer_height': 120, 'padding': 50, 'min_question_spacing': 25, 'whitespace_threshold': 128}, 'textProcessor': {'question_x_start': 100, 'question_y_start': 40, 'question_x_end': 190, 'question_y_end': 100, 'tesseract_psm_mode': 7}}, 'aqa_maths': {'imageProcessor': None, 'OCRProcessor': None}}, 'coverpage_settings': {'cambridge_science': {'regexes': {'subject': '\\b([A-Z]{6,})\\b', 'unit_code': '\\b(\\d{4})\\/\\d{2}\\b', 'component_code': '\\b\\d{4}\\/(\\d{2})\\b', 'month': '\\b([A-Z]{1}[a-z]+)\\/?[A-Z]{1}[a-z]+\\s20\\d{2}\\b', 'year': '\\bUCLES\\s+(\\d{4})\\b', 'answer': '(MARK SCHEME)'}}, 'example': {'regexes': {'subject': None, 'unit_code': None, 'component_code': None, 'month': None, 'year': None, 'type': None}}}}
2023-09-24 00:29:59,732 - exam_extractor - DEBUG - Valid formats: ['cambridge_science', 'aqa_maths']
2023-09-24 00:29:59,744 - exam_extractor - INFO - Using database: C:/Users/Tariq/Documents/birkbeck-project/database/dbOne.db
2023-09-24 00:29:59,744 - extraction_engine.managers.exam_manager - DEBUG - Setting PDFManager
2023-09-24 00:29:59,744 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 00:29:59,744 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, coverpage_settings
2023-09-24 00:29:59,744 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctcoverpage_settings, efcambridge_science
2023-09-24 00:29:59,745 - extraction_engine.managers.exam_manager - DEBUG - Setting QuestionManager
2023-09-24 00:29:59,745 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 00:29:59,745 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, exam_formats
2023-09-24 00:29:59,745 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctexam_formats, efcambridge_science
2023-09-24 00:29:59,745 - extraction_engine.managers.exam_manager - DEBUG - Setting AnswerManager
2023-09-24 00:29:59,746 - extraction_engine.managers.exam_manager - INFO - Processing exam: cambridge_science
2023-09-24 00:29:59,746 - extraction_engine.managers.exam_manager - INFO - Extracting data from PDFs...
2023-09-24 00:30:01,983 - extraction_engine.managers.exam_manager - INFO - Exam data: {'subject': 'PHYSICS', 'unit_code': '0625', 'component_code': '11', 'month': 'May', 'year': '2016', 'answer': None}
2023-09-24 00:30:01,994 - __main__ - ERROR - Mapped instance expected for relationship comparison to object.   Classes, queries and other SQL elements are not accepted in this context; for comparison with a subquery, use Exam.subject.has(**criteria).
Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\main.py", line 17, in <module>
    run_exam_extraction(exam_format, 'phys-062511-may2016.pdf', 'phys-062511-may2016-ms.pdf')
  File "C:\Users\Tariq\Documents\birkbeck-project\exam_extractor.py", line 30, in run_exam_extraction
    exam_manager.process()
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 70, in process
    raise e
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 61, in process
    exam = self._get_or_create_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 31, in _get_or_create_exam
    exam = db_session.query(Exam).filter_by(**exam_data).first()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\query.py", line 1998, in filter_by
    clauses = [
              ^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\query.py", line 1999, in <listcomp>
    _entity_namespace_key(from_entity, key) == value
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\sql\operators.py", line 582, in __eq__
    return self.operate(eq, other)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\attributes.py", line 452, in operate
    return op(self.comparator, *other, **kwargs)  # type: ignore[return-value,no-any-return]  # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\relationships.py", line 757, in __eq__
    self.property._optimized_compare(
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\relationships.py", line 1180, in _optimized_compare
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Mapped instance expected for relationship comparison to object.   Classes, queries and other SQL elements are not accepted in this context; for comparison with a subquery, use Exam.subject.has(**criteria).
2023-09-24 01:33:51,177 - __main__ - INFO - Starting exam extraction...
2023-09-24 01:33:51,177 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:33:51,177 - root - DEBUG - ConfigManager.__new__ config_path: C:\Users\Tariq\Documents\birkbeck-project\config\config.yaml
2023-09-24 01:33:51,181 - root - DEBUG - self._config in ConfigManager: {'exam_formats': {'cambridge_science': {'imageProcessor': {'binary_threshold': 180, 'margin_start': 0, 'margin_end': 180, 'question_x_start': '', 'question_x_end': '', 'question_y_start': '', 'footer_height': 120, 'padding': 50, 'min_question_spacing': 25, 'whitespace_threshold': 128}, 'textProcessor': {'question_x_start': 100, 'question_y_start': 40, 'question_x_end': 190, 'question_y_end': 100, 'tesseract_psm_mode': 7}}, 'aqa_maths': {'imageProcessor': None, 'OCRProcessor': None}}, 'coverpage_settings': {'cambridge_science': {'regexes': {'subject': '\\b([A-Z]{6,})\\b', 'unit_code': '\\b(\\d{4})\\/\\d{2}\\b', 'component_code': '\\b\\d{4}\\/(\\d{2})\\b', 'month': '\\b([A-Z]{1}[a-z]+)\\/?[A-Z]{1}[a-z]+\\s20\\d{2}\\b', 'year': '\\bUCLES\\s+(\\d{4})\\b', 'answer': '(MARK SCHEME)'}}, 'example': {'regexes': {'subject': None, 'unit_code': None, 'component_code': None, 'month': None, 'year': None, 'type': None}}}}
2023-09-24 01:33:51,181 - exam_extractor - DEBUG - Valid formats: ['cambridge_science', 'aqa_maths']
2023-09-24 01:33:51,192 - exam_extractor - INFO - Using database: C:/Users/Tariq/Documents/birkbeck-project/database/dbOne.db
2023-09-24 01:33:51,192 - extraction_engine.managers.exam_manager - DEBUG - Setting PDFManager
2023-09-24 01:33:51,193 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:33:51,193 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, coverpage_settings
2023-09-24 01:33:51,193 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctcoverpage_settings, efcambridge_science
2023-09-24 01:33:51,193 - extraction_engine.managers.exam_manager - DEBUG - Setting QuestionManager
2023-09-24 01:33:51,193 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:33:51,193 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, exam_formats
2023-09-24 01:33:51,193 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctexam_formats, efcambridge_science
2023-09-24 01:33:51,193 - extraction_engine.managers.exam_manager - DEBUG - Setting AnswerManager
2023-09-24 01:33:51,194 - extraction_engine.managers.exam_manager - INFO - Processing exam: cambridge_science
2023-09-24 01:33:51,194 - extraction_engine.managers.exam_manager - INFO - Extracting data from PDFs...
2023-09-24 01:33:53,246 - extraction_engine.managers.exam_manager - INFO - Exam data: {'subject': 'PHYSICS', 'unit_code': '0625', 'component_code': '11', 'month': 'May', 'year': '2016', 'answer': None}
2023-09-24 01:33:53,266 - __main__ - ERROR - 'CambridgeScienceImageProcessor' object has no attribute '_binary_threshold'
Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\main.py", line 17, in <module>
    run_exam_extraction(exam_format, 'phys-062511-may2016.pdf', 'phys-062511-may2016-ms.pdf')
  File "C:\Users\Tariq\Documents\birkbeck-project\exam_extractor.py", line 30, in run_exam_extraction
    exam_manager.process()
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 76, in process
    raise e
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 68, in process
    questions = self._process_questions(db_session, exam, raw_questions)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 50, in _process_questions
    return self.question_manager.execute(db_session, exam, questions_images)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\question_manager.py", line 85, in execute
    question_nums, processed_images = self._process_images(images)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\question_manager.py", line 72, in _process_images
    questions.extend(self._extract_questions(image))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\question_manager.py", line 45, in _extract_questions
    return self.image_processor.extract(image)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\processing\CambridgeScienceImageProcessor.py", line 71, in extract
    binary = self._convert_to_binary_image(grayscale, self._binary_threshold)
                                                      ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'CambridgeScienceImageProcessor' object has no attribute '_binary_threshold'
2023-09-24 01:51:51,174 - __main__ - INFO - Starting exam extraction...
2023-09-24 01:51:51,175 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:51:51,175 - root - DEBUG - ConfigManager.__new__ config_path: C:\Users\Tariq\Documents\birkbeck-project\config\config.yaml
2023-09-24 01:51:51,179 - root - DEBUG - self._config in ConfigManager: {'exam_formats': {'cambridge_science': {'imageProcessor': {'binary_threshold': 180, 'margin_start': 0, 'margin_end': 180, 'question_x_start': '', 'question_x_end': '', 'question_y_start': '', 'footer_height': 120, 'padding': 50, 'min_question_spacing': 25, 'whitespace_threshold': 128}, 'textProcessor': {'question_x_start': 100, 'question_y_start': 40, 'question_x_end': 190, 'question_y_end': 100, 'tesseract_psm_mode': 7}}, 'aqa_maths': {'imageProcessor': None, 'OCRProcessor': None}}, 'coverpage_settings': {'cambridge_science': {'regexes': {'subject': '\\b([A-Z]{6,})\\b', 'unit_code': '\\b(\\d{4})\\/\\d{2}\\b', 'component_code': '\\b\\d{4}\\/(\\d{2})\\b', 'month': '\\b([A-Z]{1}[a-z]+)\\/?[A-Z]{1}[a-z]+\\s20\\d{2}\\b', 'year': '\\bUCLES\\s+(\\d{4})\\b', 'answer': '(MARK SCHEME)'}}, 'example': {'regexes': {'subject': None, 'unit_code': None, 'component_code': None, 'month': None, 'year': None, 'type': None}}}}
2023-09-24 01:51:51,179 - exam_extractor - DEBUG - Valid formats: ['cambridge_science', 'aqa_maths']
2023-09-24 01:51:51,190 - exam_extractor - INFO - Using database: C:/Users/Tariq/Documents/birkbeck-project/database/dbOne.db
2023-09-24 01:51:51,190 - extraction_engine.managers.exam_manager - DEBUG - Setting PDFManager
2023-09-24 01:51:51,190 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:51:51,190 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, coverpage_settings
2023-09-24 01:51:51,190 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctcoverpage_settings, efcambridge_science
2023-09-24 01:51:51,190 - extraction_engine.managers.exam_manager - DEBUG - Setting QuestionManager
2023-09-24 01:51:51,190 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:51:51,191 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, exam_formats
2023-09-24 01:51:51,191 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctexam_formats, efcambridge_science
2023-09-24 01:51:51,191 - extraction_engine.managers.exam_manager - DEBUG - Setting AnswerManager
2023-09-24 01:51:51,191 - extraction_engine.managers.exam_manager - INFO - Processing exam: cambridge_science
2023-09-24 01:51:51,192 - extraction_engine.managers.exam_manager - INFO - Extracting data from PDFs...
2023-09-24 01:51:53,257 - extraction_engine.managers.exam_manager - INFO - Exam data: {'subject': 'PHYSICS', 'unit_code': '0625', 'component_code': '11', 'month': 'May', 'year': '2016', 'answer': None}
2023-09-24 01:51:53,266 - __main__ - ERROR - Query.filter() got an unexpected keyword argument 'month'
Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\main.py", line 17, in <module>
    run_exam_extraction(exam_format, 'phys-062511-may2016.pdf', 'phys-062511-may2016-ms.pdf')
  File "C:\Users\Tariq\Documents\birkbeck-project\exam_extractor.py", line 30, in run_exam_extraction
    exam_manager.process()
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 73, in process
    raise e
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 63, in process
    exam = self._get_or_create_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 32, in _get_or_create_exam
    exam = get_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\database\database_utils.py", line 46, in get_exam
    exam = session.query(Exam).filter(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Query.filter() got an unexpected keyword argument 'month'
2023-09-24 01:55:49,295 - __main__ - INFO - Starting exam extraction...
2023-09-24 01:55:49,295 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:55:49,296 - root - DEBUG - ConfigManager.__new__ config_path: C:\Users\Tariq\Documents\birkbeck-project\config\config.yaml
2023-09-24 01:55:49,300 - root - DEBUG - self._config in ConfigManager: {'exam_formats': {'cambridge_science': {'imageProcessor': {'binary_threshold': 180, 'margin_start': 0, 'margin_end': 180, 'question_x_start': '', 'question_x_end': '', 'question_y_start': '', 'footer_height': 120, 'padding': 50, 'min_question_spacing': 25, 'whitespace_threshold': 128}, 'textProcessor': {'question_x_start': 100, 'question_y_start': 40, 'question_x_end': 190, 'question_y_end': 100, 'tesseract_psm_mode': 7}}, 'aqa_maths': {'imageProcessor': None, 'OCRProcessor': None}}, 'coverpage_settings': {'cambridge_science': {'regexes': {'subject': '\\b([A-Z]{6,})\\b', 'unit_code': '\\b(\\d{4})\\/\\d{2}\\b', 'component_code': '\\b\\d{4}\\/(\\d{2})\\b', 'month': '\\b([A-Z]{1}[a-z]+)\\/?[A-Z]{1}[a-z]+\\s20\\d{2}\\b', 'year': '\\bUCLES\\s+(\\d{4})\\b', 'answer': '(MARK SCHEME)'}}, 'example': {'regexes': {'subject': None, 'unit_code': None, 'component_code': None, 'month': None, 'year': None, 'type': None}}}}
2023-09-24 01:55:49,300 - exam_extractor - DEBUG - Valid formats: ['cambridge_science', 'aqa_maths']
2023-09-24 01:55:49,312 - exam_extractor - INFO - Using database: C:/Users/Tariq/Documents/birkbeck-project/database/dbOne.db
2023-09-24 01:55:49,312 - extraction_engine.managers.exam_manager - DEBUG - Setting PDFManager
2023-09-24 01:55:49,312 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:55:49,312 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, coverpage_settings
2023-09-24 01:55:49,312 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctcoverpage_settings, efcambridge_science
2023-09-24 01:55:49,312 - extraction_engine.managers.exam_manager - DEBUG - Setting QuestionManager
2023-09-24 01:55:49,312 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 01:55:49,313 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, exam_formats
2023-09-24 01:55:49,313 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctexam_formats, efcambridge_science
2023-09-24 01:55:49,313 - extraction_engine.managers.exam_manager - DEBUG - Setting AnswerManager
2023-09-24 01:55:49,313 - extraction_engine.managers.exam_manager - INFO - Processing exam: cambridge_science
2023-09-24 01:55:49,313 - extraction_engine.managers.exam_manager - INFO - Extracting data from PDFs...
2023-09-24 01:55:51,383 - extraction_engine.managers.exam_manager - INFO - Exam data: {'subject': 'PHYSICS', 'unit_code': '0625', 'component_code': '11', 'month': 'May', 'year': '2016', 'answer': None}
2023-09-24 01:55:51,396 - __main__ - ERROR - (sqlite3.OperationalError) no such table: exams
[SQL: SELECT exams.id AS exams_id, exams.exam_board AS exams_exam_board, exams.month AS exams_month, exams.year AS exams_year, exams.unit_code AS exams_unit_code, exams.component_code AS exams_component_code, exams.subject_id AS exams_subject_id 
FROM exams 
WHERE exams.month = ? AND exams.year = ? AND exams.unit_code = ? AND exams.component_code = ?
 LIMIT ? OFFSET ?]
[parameters: ('May', '2016', '0625', '11', 1, 0)]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such table: exams

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\main.py", line 17, in <module>
    run_exam_extraction(exam_format, 'phys-062511-may2016.pdf', 'phys-062511-may2016-ms.pdf')
  File "C:\Users\Tariq\Documents\birkbeck-project\exam_extractor.py", line 30, in run_exam_extraction
    exam_manager.process()
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 73, in process
    raise e
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 63, in process
    exam = self._get_or_create_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 32, in _get_or_create_exam
    exam = get_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\database\database_utils.py", line 51, in get_exam
    ).first()
      ^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\query.py", line 2743, in first
    return self.limit(1)._iter().first()  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\query.py", line 2842, in _iter
    result: Union[ScalarResult[_T], Result[_T]] = self.session.execute(
                                                  ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\session.py", line 2262, in execute
    return self._execute_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\session.py", line 2144, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\orm\context.py", line 293, in orm_execute_statement
    result = conn.execute(
             ^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1412, in execute
    return meth(
           ^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\sql\elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1635, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1844, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 2339, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "C:\Users\Tariq\Documents\birkbeck-project\project-env\Lib\site-packages\sqlalchemy\engine\default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: exams
[SQL: SELECT exams.id AS exams_id, exams.exam_board AS exams_exam_board, exams.month AS exams_month, exams.year AS exams_year, exams.unit_code AS exams_unit_code, exams.component_code AS exams_component_code, exams.subject_id AS exams_subject_id 
FROM exams 
WHERE exams.month = ? AND exams.year = ? AND exams.unit_code = ? AND exams.component_code = ?
 LIMIT ? OFFSET ?]
[parameters: ('May', '2016', '0625', '11', 1, 0)]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
2023-09-24 02:04:49,952 - __main__ - INFO - Starting exam extraction...
2023-09-24 02:04:49,999 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 02:04:49,999 - root - DEBUG - ConfigManager.__new__ config_path: C:\Users\Tariq\Documents\birkbeck-project\config\config.yaml
2023-09-24 02:04:50,003 - root - DEBUG - self._config in ConfigManager: {'exam_formats': {'cambridge_science': {'imageProcessor': {'binary_threshold': 180, 'margin_start': 0, 'margin_end': 180, 'question_x_start': '', 'question_x_end': '', 'question_y_start': '', 'footer_height': 120, 'padding': 50, 'min_question_spacing': 25, 'whitespace_threshold': 128}, 'textProcessor': {'question_x_start': 100, 'question_y_start': 40, 'question_x_end': 190, 'question_y_end': 100, 'tesseract_psm_mode': 7}}, 'aqa_maths': {'imageProcessor': None, 'OCRProcessor': None}}, 'coverpage_settings': {'cambridge_science': {'regexes': {'subject': '\\b([A-Z]{6,})\\b', 'unit_code': '\\b(\\d{4})\\/\\d{2}\\b', 'component_code': '\\b\\d{4}\\/(\\d{2})\\b', 'month': '\\b([A-Z]{1}[a-z]+)\\/?[A-Z]{1}[a-z]+\\s20\\d{2}\\b', 'year': '\\bUCLES\\s+(\\d{4})\\b', 'answer': '(MARK SCHEME)'}}, 'example': {'regexes': {'subject': None, 'unit_code': None, 'component_code': None, 'month': None, 'year': None, 'type': None}}}}
2023-09-24 02:04:50,003 - exam_extractor - DEBUG - Valid formats: ['cambridge_science', 'aqa_maths']
2023-09-24 02:04:50,003 - exam_extractor - INFO - Using database: C:/Users/Tariq/Documents/birkbeck-project/database/dbOne.db
2023-09-24 02:04:50,003 - extraction_engine.managers.exam_manager - DEBUG - Setting PDFManager
2023-09-24 02:04:50,003 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 02:04:50,004 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, coverpage_settings
2023-09-24 02:04:50,004 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctcoverpage_settings, efcambridge_science
2023-09-24 02:04:50,004 - extraction_engine.managers.exam_manager - DEBUG - Setting QuestionManager
2023-09-24 02:04:50,004 - root - DEBUG - ConfigManager.__new__ called
2023-09-24 02:04:50,004 - extraction_engine.managers.config_manager - DEBUG - ConfigManager.get_config called with ef: cambridge_science, exam_formats
2023-09-24 02:04:50,004 - extraction_engine.managers.config_manager - DEBUG - inside 3rd if: ctexam_formats, efcambridge_science
2023-09-24 02:04:50,004 - extraction_engine.managers.exam_manager - DEBUG - Setting AnswerManager
2023-09-24 02:04:50,005 - extraction_engine.managers.exam_manager - INFO - Processing exam: cambridge_science
2023-09-24 02:04:50,005 - extraction_engine.managers.exam_manager - INFO - Extracting data from PDFs...
2023-09-24 02:04:52,049 - extraction_engine.managers.exam_manager - INFO - Exam data: {'subject': 'PHYSICS', 'unit_code': '0625', 'component_code': '11', 'month': 'May', 'year': '2016', 'answer': None}
2023-09-24 02:04:52,054 - __main__ - ERROR - exam_board is required for Exam creation.
Traceback (most recent call last):
  File "C:\Users\Tariq\Documents\birkbeck-project\main.py", line 18, in <module>
    run_exam_extraction(exam_format, 'phys-062511-may2016.pdf', 'phys-062511-may2016-ms.pdf')
  File "C:\Users\Tariq\Documents\birkbeck-project\exam_extractor.py", line 30, in run_exam_extraction
    exam_manager.process()
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 73, in process
    raise e
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 63, in process
    exam = self._get_or_create_exam(db_session, exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\managers\exam_manager.py", line 36, in _get_or_create_exam
    exam = exam_factory.create_exam(exam_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\factories\exam_factory.py", line 18, in create_exam
    self._validate_exam_data(exam_data)
  File "C:\Users\Tariq\Documents\birkbeck-project\extraction_engine\factories\exam_factory.py", line 13, in _validate_exam_data
    raise ValueError(f"{field} is required for Exam creation.")
ValueError: exam_board is required for Exam creation.
